Ubuntu 20.04 python unit tests failures

running python unit test `make py-tests` on Ubuntu 20.04 (Focal) yields to the following three errors:

**Running ganeti.ovf_unittest.py**
```text
.............................................................FF.....F
======================================================================
FAIL: testOVFWriterInit (__main__.TestOVFWriter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/py/ganeti.ovf_unittest.py", line 794, in testOVFWriterInit
    self.assertTrue(EXPORT_EMPTY in result)
AssertionError: False is not true

======================================================================
FAIL: testSaveDisksData (__main__.TestOVFWriter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/py/ganeti.ovf_unittest.py", line 804, in testSaveDisksData
    self.assertTrue(EXPORT_DISKS in result)
AssertionError: False is not true

======================================================================
FAIL: testSaveVirtualSystem (__main__.TestOVFWriter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/py/ganeti.ovf_unittest.py", line 831, in testSaveVirtualSystem
    self.assertTrue(EXPORT_SYSTEM in result)
AssertionError: False is not true

----------------------------------------------------------------------
Ran 69 tests in 0.124s

FAILED (failures=3)
FAIL: test/py/ganeti.ovf_unittest.py
```
**Running ganeti.objects_unittest.py**
```text
........................E...............
======================================================================
ERROR: testCheckISpecSyntax (__main__.TestInstancePolicy)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/py/ganeti.objects_unittest.py", line 546, in testCheckISpecSyntax
    for param in spec:
RuntimeError: dictionary keys changed during iteration

----------------------------------------------------------------------
Ran 40 tests in 0.012s

FAILED (errors=1)
FAIL: test/py/ganeti.objects_unittest.py
```

**Running ganeti.cli_unittest.py**
```text
........./test/py/ganeti.cli_unittest.py:1151: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.
  return yaml.load(buf.getvalue())
E................................................................
======================================================================
ERROR: testWithDefaults (__main__.TestFormatPolicyInfo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./test/py/ganeti.cli_unittest.py", line 1182, in testWithDefaults
    self._PrintAndCheckParsed(constants.IPOLICY_DEFAULTS)
  File "./test/py/ganeti.cli_unittest.py", line 1155, in _PrintAndCheckParsed
    self._TranslateParsedNames(parsed)
  File "./test/py/ganeti.cli_unittest.py", line 1138, in _TranslateParsedNames
    for key in minmax:
RuntimeError: dictionary keys changed during iteration

----------------------------------------------------------------------
Ran 73 tests in 0.019s

FAILED (errors=1)
FAIL: test/py/ganeti.cli_unittest.py
```

--- a/test/py/ganeti.cli_unittest.py
+++ b/test/py/ganeti.cli_unittest.py
@@ -1135,7 +1135,7 @@ class TestFormatPolicyInfo(unittest.TestCase):
       ]:
       self._RenameDictItem(parsed, pretty, raw)
     for minmax in parsed[constants.ISPECS_MINMAX]:
-      for key in minmax:
+      for key in set(minmax.keys()):
         keyparts = key.split("/", 1)
         if len(keyparts) > 1:
           self._RenameDictItem(minmax, key, keyparts[0])

--- a/test/py/ganeti.objects_unittest.py
+++ b/test/py/ganeti.objects_unittest.py
@@ -543,7 +543,7 @@ class TestInstancePolicy(unittest.TestCase):
     bad_ipolicy = copy.deepcopy(good_ipolicy)
     for minmax in bad_ipolicy[constants.ISPECS_MINMAX]:
       for (key, spec) in minmax.items():
-        for param in spec:
+        for param in set(spec.keys()):
           oldv = spec[param]
           del spec[param]
           self._AssertPolicyIsBad(bad_ipolicy)
@@ -554,7 +554,7 @@ class TestInstancePolicy(unittest.TestCase):
     assert bad_ipolicy == good_ipolicy
 
     stdspec = bad_ipolicy[constants.ISPECS_STD]
-    for param in stdspec:
+    for param in set(stdspec.keys()):
       oldv = stdspec[param]
       del stdspec[param]
       self._AssertPolicyIsBad(bad_ipolicy, True)

--- a/test/py/ganeti.ovf_unittest.py
+++ b/test/py/ganeti.ovf_unittest.py
@@ -347,6 +347,26 @@ def _GetArgs(args, with_name=False):
   return options
 
 
+def _SortAttrs(tree):
+  """Sort XML node attributes by name
+
+  As of Python 3.8, ElementTree will sort output according to insertion order,
+  leveraging the stable dict sorting available since Python 3.6. Some of the
+  tests in this file rely on the previous behavior of ElementTree, which
+  produced stable output by sorting the node attributes.
+
+  This function updates node attribute dictionaries so that the insertion order
+  matches the alphabetical order and the tree thus produces identical output in
+  all Python 3 versions.
+  """
+  for el in tree.iter():
+    attrib = el.attrib
+    if len(attrib) > 1:
+      attribs = sorted(attrib.items())
+      attrib.clear()
+      attrib.update(attribs)
+
+
 OPTS_EMPTY = _GetArgs(ARGS_EMPTY)
 OPTS_EXPORT_NO_NAME = _GetArgs(ARGS_EXPORT_DIR)
 OPTS_EXPORT = _GetArgs(ARGS_EXPORT_DIR, with_name=True)
@@ -789,46 +809,43 @@ class TestOVFWriter(BetterUnitTest):
   def tearDown(self):
     pass
 
-  def testOVFWriterInit(self):
+  def assertXMLOutputContains(self, text):
+    _SortAttrs(self.writer.tree)
     result = ET.tostring(self.writer.tree, encoding="unicode")
-    self.assertTrue(EXPORT_EMPTY in result)
+    self.assertIn(text, result)
+
+  def testOVFWriterInit(self):
+    self.assertXMLOutputContains(EXPORT_EMPTY)
 
   def testSaveDisksDataEmpty(self):
     self.writer.SaveDisksData([])
-    result = ET.tostring(self.writer.tree, encoding="unicode")
-    self.assertTrue(EXPORT_DISKS_EMPTY in result)
+    self.assertXMLOutputContains(EXPORT_DISKS_EMPTY)
 
   def testSaveDisksData(self):
     self.writer.SaveDisksData(EXP_DISKS_LIST)
-    result = ET.tostring(self.writer.tree, encoding="unicode")
-    self.assertTrue(EXPORT_DISKS in result)
+    self.assertXMLOutputContains(EXPORT_DISKS)
 
   def testSaveNetworkDataEmpty(self):
     self.writer.SaveNetworksData([])
-    result = ET.tostring(self.writer.tree, encoding="unicode")
-    self.assertTrue(EXPORT_NETWORKS_EMPTY in result)
+    self.assertXMLOutputContains(EXPORT_NETWORKS_EMPTY)
 
   def testSaveNetworksData(self):
     self.writer.SaveNetworksData(EXP_NETWORKS_LIST)
-    result = ET.tostring(self.writer.tree, encoding="unicode")
-    self.assertTrue(EXPORT_NETWORKS in result)
+    self.assertXMLOutputContains(EXPORT_NETWORKS)
 
   def testSaveGanetiDataIncomplete(self):
     self.writer.SaveGanetiData(EXP_PARTIAL_GANETI_DICT, EXP_NETWORKS_LIST)
-    result = ET.tostring(self.writer.tree, encoding="unicode")
-    self.assertTrue(EXPORT_GANETI_INCOMPLETE in result)
+    self.assertXMLOutputContains(EXPORT_GANETI_INCOMPLETE)
 
   def testSaveGanetiDataComplete(self):
     self.writer.SaveGanetiData(EXP_GANETI_DICT, EXP_NETWORKS_LIST)
-    result = ET.tostring(self.writer.tree, encoding="unicode")
-    self.assertTrue(EXPORT_GANETI in result)
+    self.assertXMLOutputContains(EXPORT_GANETI)
 
   def testSaveVirtualSystem(self):
     self.writer.SaveDisksData(EXP_DISKS_LIST)
     self.writer.SaveNetworksData(EXP_NETWORKS_LIST)
     self.writer.SaveVirtualSystemData(EXP_NAME, EXP_VCPUS, EXP_MEMORY)
-    result = ET.tostring(self.writer.tree, encoding="unicode")
-    self.assertTrue(EXPORT_SYSTEM in result)
+    self.assertXMLOutputContains(EXPORT_SYSTEM)
 
 
 if __name__ == "__main__":

--- a/test/py/ganeti.serializer_unittest.py
+++ b/test/py/ganeti.serializer_unittest.py
@@ -202,11 +202,11 @@ class TestPrivate(unittest.TestCase):
     pDict = serializer.PrivateDict()
     pDict["bar"] = "egg"
 
-    self.assertTrue("egg" is pDict["bar"].Get(),
+    self.assertTrue("egg" == pDict["bar"].Get(),
                     "Value not returned by Private.Get()")
-    self.assertTrue("egg" is pDict.GetPrivate("bar"),
+    self.assertTrue("egg" == pDict.GetPrivate("bar"),
                     "Value not returned by Private.GetPrivate()")
-    self.assertTrue("egg" is pDict.Unprivate()["bar"],
+    self.assertTrue("egg" == pDict.Unprivate()["bar"],
                     "Value not returned by PrivateDict.Unprivate()")
 
     json = serializer.Dump(pDict,
@@ -215,7 +215,7 @@ class TestPrivate(unittest.TestCase):
 
   def testDictGet(self):
     result = serializer.PrivateDict().GetPrivate("bar", "tar")
-    self.assertTrue("tar" is result,
+    self.assertTrue("tar" == result,
                     "Private.GetPrivate() did not handle the default case")
 
   def testZeronessPrivate(self):

